### DISCUSSION TITLE
Generating 1D lower dimensional block and adding it to 3D filemesh

### DISCUSSION THREAD

**Discussion by Florian-Konrad** [idaholab/moose#24887] [Q&A Meshing] [answered]
Hi :)
I was wondering if there is a way with the existing MeshGenerators to generate a lower dimensional block from a list of points (x,y,z) that describe a polyline and add this to an existing 3D filemesh. The polyline in my case is also following the edges of the filemesh. 
My question is similar to this one https://github.com/idaholab/moose/discussions/24174
But my filemesh (exodus format) is not coming from GMSH.
If this isnt possible does someone know a way/tool to do this?
Cheers
Florian

**Comment by GiudGiud** [comment, marked as answer]
Hello

I m not sure we have tried to do this before but on paper it's possible.
Use a FileMeshGenerator to load your mesh
Use a ParsedCurveGenerator to create the 1D line (for example)

then you can either combine (detached meshes) using a CombineMeshGenerator or stitch them (you ll need a nodeset to stitch the polyline to on the multiD mesh)

Guillaume

**Comment by Florian-Konrad** [comment]
Ah I see! Thanks a lot! ... maybe this did the trick.. need to test if the mesh actually works for my simulation but in paraview it looks good!
Did it like this:
```
[Mesh]
  [input_mesh]
    type = FileMeshGenerator
    file = ./test_mesh.e
  []
  [wellblock]
    type = PolyLineMeshGenerator
    points = '0.2	0.5	0
    0.2	0.5	0.0777778
    0.2	0.5	0.155556
    0.2	0.5	0.233333
    0.2	0.5	0.311111
    0.2	0.5	0.388889
    0.2	0.5	0.466667
    0.2	0.5	0.544444
    0.2	0.5	0.622222
    0.2	0.5	0.7    
    '
    loop = false
  []
  [wellblockbc]
    type = ExtraNodesetGenerator
    new_boundary = 'well_block_bc'
    coord = '0.2	0.5	0;
    0.2	0.5	0.0777778;
    0.2	0.5	0.155556;
    0.2	0.5	0.233333;
    0.2	0.5	0.311111;
    0.2	0.5	0.388889;
    0.2	0.5	0.466667;
    0.2	0.5	0.544444;
    0.2	0.5	0.622222;
    0.2	0.5	0.7    
    '
    input = wellblock
  []
  [aquiferblockwithwellbc]
    type = ExtraNodesetGenerator
    new_boundary = 'well_bc_aq'
    coord = '0.2	0.5	0;
    0.2	0.5	0.0777778;
    0.2	0.5	0.155556;
    0.2	0.5	0.233333;
    0.2	0.5	0.311111;
    0.2	0.5	0.388889;
    0.2	0.5	0.466667;
    0.2	0.5	0.544444;
    0.2	0.5	0.622222;
    0.2	0.5	0.7    
    '
    input = input_mesh
  []
  [stitch]
    type = StitchedMeshGenerator
    inputs = 'aquiferblockwithwellbc wellblockbc'
    clear_stitched_boundary_ids = true
    stitch_boundaries_pairs = 'well_bc_aq well_block_bc'
  []
[]
```

**Reply by GiudGiud** [reply]
The stitcher is verbose. Did it tell you the stitching succeeded for at least one node?

**Reply by Florian-Konrad** [reply]
This is the output, looks good right?

```
mpiexec -n 4 ~/projects/golem/golem-opt -i generatedmeshtests.i --mesh-only
In UnstructuredMesh::stitch_meshes:
This mesh has 10 nodes on boundary 1.
Other mesh has 10 nodes on boundary 3.
Minimum edge length on both surfaces is 0.071527.
In UnstructuredMesh::stitch_meshes:
Found 10 matching nodes.

 Mesh Information:
  elem_dimensions()={1, 3}
  spatial_dimension()=3
  n_nodes()=2170
    n_local_nodes()=623
  n_elem()=10431
    n_local_elem()=2606
    n_active_elem()=10431
  n_subdomains()=2
  n_elemsets()=0
  n_partitions()=4
  n_processors()=4
  n_threads()=1
  processor_id()=0
  is_prepared()=true
  is_replicated()=true

 Mesh Bounding Box:
  Minimum: (x,y,z)=(       0,     -0.2,        0)
  Maximum: (x,y,z)=(     1.1,      1.2,      0.7)
  Delta:   (x,y,z)=(     1.1,      1.4,      0.7)

 Mesh Element Type(s):
  EDGE2, TET4

 Mesh Nodesets:
  Nodeset 1 (well_bc_aq), 10 nodes
   Bounding box minimum: (x,y,z)=(     0.2,      0.5,        0)
   Bounding box maximum: (x,y,z)=(-1.79769e+308, -1.79769e+308, -1.79769e+308)
   Bounding box delta: (x,y,z)=(-1.79769e+308, -1.79769e+308, -1.79769e+308)
  Nodeset 2, 1 nodes
   Bounding box minimum: (x,y,z)=(     0.2,      0.5,      0.7)
   Bounding box maximum: (x,y,z)=(-1.79769e+308, -1.79769e+308, -1.79769e+308)
   Bounding box delta: (x,y,z)=(-1.79769e+308, -1.79769e+308, -1.79769e+308)
  Nodeset 3 (well_block_bc), 10 nodes
   Bounding box minimum: (x,y,z)=(     0.2,      0.5,        0)
   Bounding box maximum: (x,y,z)=(-1.79769e+308, -1.79769e+308, -1.79769e+308)
   Bounding box delta: (x,y,z)=(-1.79769e+308, -1.79769e+308, -1.79769e+308)

 Mesh Sidesets:
  Sideset 1, 1 sides (NODEELEM), 1 elems (EDGE2), 1 nodes
   Side volume: 1
   Bounding box minimum: (x,y,z)=(     0.2,      0.5,        0)
   Bounding box maximum: (x,y,z)=(     0.2,      0.5,        0)
   Bounding box delta: (x,y,z)=(       0,        0,        0)
  Sideset 2, 1 sides (NODEELEM), 1 elems (EDGE2), 1 nodes
   Side volume: 1
   Bounding box minimum: (x,y,z)=(     0.2,      0.5,      0.7)
   Bounding box maximum: (x,y,z)=(     0.2,      0.5,      0.7)
   Bounding box delta: (x,y,z)=(       0,        0,        0)

 Mesh Edgesets:
  None

 Mesh Subdomains:
  Subdomain 0: 9 elems (EDGE2, 9 active), 10 active nodes
   Volume: 0.7
   Bounding box minimum: (x,y,z)=(     0.2,      0.5,        0)
   Bounding box maximum: (x,y,z)=(     0.2,      0.5,      0.7)
   Bounding box delta: (x,y,z)=(       0,        0,      0.7)
  Subdomain 2: 10422 elems (TET4, 10422 active), 2170 active nodes
   Volume: 0.7245
   Bounding box minimum: (x,y,z)=(       0,     -0.2,        0)
   Bounding box maximum: (x,y,z)=(     1.1,      1.2,      0.7)
   Bounding box delta: (x,y,z)=(     1.1,      1.4,      0.7)
  Global mesh volume = 1.4245

```

**Reply by GiudGiud** [reply]
looks fine but you are stitiching all 10 nodes of the 1D mesh? So the nodeset and the 1D mesh matched exactly?

If you were just trying to create a block at a sideset location, there's this mesh generator you can use:
https://mooseframework.inl.gov/source/meshgenerators/LowerDBlockFromSidesetGenerator.html

**Reply by Florian-Konrad** [reply]
Jeah so I got the nodeset by extracting it from my existing mesh. I want to set a 1D mesh on the edges of my existing mesh, so yes they match exactly. 
I played around with LowerDBlockFromSidesetGenerator but it didnt work, if I understand correctly it tries to generate a 2D block if I input a 3D mesh. And I wanted to only have the edges. 
And the new block should be connected to the 3D mesh along all nodes it shares.
For more context, in a hydraulic simulation (Darcy, porous flow, GOLEM application containing the physics) the 1D block should represent a highly conductive structure (a well) that can exchange fluid flow with the surrounding 3D-mesh-elements. So if I understand it correctly the block needs be connected on all shared nodes.

**Reply by Florian-Konrad** [reply]
Here you can see the result with a test mesh I used to figure out the workflow and how I can do this later on with a different and more complicted one:
![stitch_test ong](https://github.com/idaholab/moose/assets/72145069/bb40804f-b95a-4f4b-86c0-11d2d64d1e67)

**Reply by GiudGiud** [reply]
sounds good for the simple mesh. You ll want to make sure you get physical and correct solutions on that one too (not just a working mesh) before moving to something complicated.

I think iterating on "lowerD" could have gotten you from 3D to 2D to 1D. The way you do it works

**Reply by Florian-Konrad** [reply]
ah interesting, will try all that ! Thanks a lot for the help :)

**Comment by Florian-Konrad** [comment]
Hey guys!

I have an additional question/situation regarding this stitching process.

Coming from a different filemesh (structured mesh based on an eclipse model using this python tool to convert it to exodus: https://github.com/cpgr/em2ex) I now get an error that says stitching is not possible due to the mesh type.

See attached the error, as well as the Mesh file and the input I use with --mesh-only to test the stitching.

Is it generally true that its not possible to add a lower 1D block to a structured 3D-hex-mesh? Are there any workarounds to this?
I somehow have to add a discrete element following an existing edge to this mesh!


<img width="914" alt="MicrosoftTeams-image (2)" src="https://github.com/idaholab/moose/assets/72145069/f4c056b0-caaa-408b-9421-9d21d412d911">

[MeshStitching.zip](https://github.com/idaholab/moose/files/12144290/MeshStitching.zip)


Cheers
Florian

**Reply by GiudGiud** [reply]
Hi Florian

we definitely support loading these meshes with mixed dimensional elements. 

For your error right now it s complaining because you are trying to make a block with mixed element types and that s not supported 

use a renameBlockGenerator before stitching to have a subdomain for each element type

**Reply by Florian-Konrad** [reply]
Hi Guillaume,
thanks that makes sense... and I tried it but the error doesnt change unfortunately.
See my input file attached here:
[Boxmodell_stiched_input.txt](https://github.com/idaholab/moose/files/12145906/Boxmodell_stiched_input.txt)

**Reply by GiudGiud** [reply]
if you set show_info = true in that last generator, do you get confirmation that the elements of different types are in different blocks?

**Reply by Florian-Konrad** [reply]
ok thanks for the hints.. so I played around with show_info and looked at the renaming of my blocks with Paraview and I found that for some reason the RenameBlockGenerator behaves unexpected for my filemesh. 

- renaming the filemesh to a string doesnt have any effect (eg. new_block = "aquifer" --> id = 0 and name = 0)
- renaming to an integer changes the id value shown by Paraview but not the name of the block
- from my example input below I would expect my filemesh to be renamed to "block_21" with the id 21 but as shown in the screenshot only the id is changed not the name of the block
- for the PolyLineMeshGenerator block it works as expected --> id = 4 and name = "block_4"

```
[input_mesh]
    type = FileMeshGenerator
    file = ./MESH.e
[]
[rename1]
    type = RenameBlockGenerator
    input = input_mesh
    new_block = 21
    old_block = 0
[]

[wellblock]
    type = PolyLineMeshGenerator
    points = 'some coords..'
    loop = false
[]
[rename2]
    type = RenameBlockGenerator
    input = wellblock
    new_block = 4
    old_block = 0
[]
``` 
<img width="429" alt="renamed_blocks_paraview" src="https://github.com/idaholab/moose/assets/72145069/9977d3f2-6d51-4c60-9e3b-c4081db46729">

**Reply by GiudGiud** [reply]
Thanks for the detailed report. 
it was patched to become simpler with no need to specify name/ids but it seems there’s issues

once the rename are changing the ids, do you get a correct stitch?

**Reply by Florian-Konrad** [reply]
yes as soon as the id values of my blocks are different stitching works! should I open an issue from this?

**Reply by GiudGiud** [reply]
As you prefer. Please tag me if you do

**Reply by Florian-Konrad** [reply]
I just discovered a bit more .. seems like my filemesh has one block without any real id ... but moose treats it as id = 0 
Changing the non existing id to 0 or a string has no effect. But changing it to >0 integer applies the new id properly so its possible aftwards to change it also to a string with a second RenameBlockGenerator
So this works ... not sure if this is worth an issue since it seems to come from my filemesh.
```
[Mesh]
  [input_mesh]
    type = FileMeshGenerator
    file = ./MESH.e
  []
  [rename1]
    type = RenameBlockGenerator
    input = input_mesh
    new_block = 3
    old_block = 0
  []
  [rename1_renamed]
    type = RenameBlockGenerator
    input = rename1
    new_block = 'aquifer'
    old_block = 3
  []
[]
```

**Reply by GiudGiud** [reply]
This worked:
```
[Mesh]
  [input_mesh]
    type = FileMeshGenerator
    file = ./MESH.e
  []
  [re]
    type = RenameBlockGenerator
    input = input_mesh
    old_block = '0'
    new_block = 'hhss'
  []
[]
```

it gave me the following:
```

 Mesh Subdomains:
  Subdomain 0 (hhss): 350000 elems (HEX8, 350000 active), 367236 active nodes
   Volume: 1.6e+12
   Bounding box minimum: (x,y,z)=(       0,        0,    -4000)
   Bounding box maximum: (x,y,z)=(   20000,    20000,        0)
   Bounding box delta: (x,y,z)=(   20000,    20000,     4000)
  Global mesh volume = 1.6e+12
```

so I dont think we've narrowed down the issue enough yet. 
StitchedMG does not present block id overlap between the two parts. I m not sure we want to change that since it's easy enough to Rename normally?

**Reply by Florian-Konrad** [reply]
jeah no need to change this!

**Comment by Florian-Konrad** [comment]
I have a related follow up question!
The element type generated by PolyLineMeshGenerator is EDGE2. In previous meshes i used (generated externally) my well elements (1d lower dimensional block) seem to be of type TRUSS.

Cant find information on these types just that for libmesh TRUSS elements end up being treated like EDGE2:
https://mooseframework.inl.gov/docs/doxygen/libmesh/exodusII__io__helper_8C_source.html

But I'm wondering what the difference is between EDGE2 and TRUSS for my actual simulation? Is it something related to degrees of freedom? Where can I find more info on this?

**Reply by GiudGiud** [reply]
@roystgnr will know.

**Reply by GiudGiud** [reply]
@jain651 is this something you have worked on as well?

**Reply by roystgnr** [reply]
In libMesh, an ExodusII `TRUSS` element is just an `EDGE2` with a funny birth name which is `read()` but then immediately changed and forgotten.

I'm not actually sure whether there's a difference to most other ExodusII codes either; at least https://github.com/sandialabs/seacas/blob/master/packages/seacas/libraries/exodus/include/exodus-element-types.md suggests that `TRUSS` and `BAR` (and `TRUSS2` and `BAR2`) are indistinguishable to them.

In older finite element libraries you'll often see a lack of factorization of "finite element family" vs "geometric element" vs "weak form", and AFAIK that's where a lot of these naming distinctions come from.  In libMesh and MOOSE you can take e.g. an Edge2 and give it one Lagrange or Hierarchic variable for continuous Galerkin heat conduction or one Hermite variable for conduction or phase field evolution or three Lagrange variables for displacement or multi-species conduction or 6 for displacements and moments or whatever, and it's still an Edge2, but in some other codes the "element type" refers to each unique combination of geometry *and* discretization (including degree) *and* physics.

**Reply by Florian-Konrad** [reply]
ok so if I understand correctly it shouldn't matter for any moose app?
as libmesh is only translating all synonyms to EDGE2 and there is no mechanism that limits anything on EDGE2 elements depending on their "funny birthname".

**Reply by roystgnr** [reply]
It can't matter to any app reading files only via the standard MOOSE APIs - there's not even any information retained by libMesh or MOOSE classes to tell which original name got translated to `EDGE2`.  MOOSE apps do have direct access to ExodusII functions, so it's not impossible for some app developer to re-open a file *themselves* with code that executes differently depending on what they find, but surely they wouldn't...

**Reply by Florian-Konrad** [reply]
perfect thanks for explaining!
